src-java/kilda-persistence-tinkerpop/src/main/java/org/openkilda/persistence/ferma/repositories/FermaFlowPathRepository.java

```diff
@@ -78,6 +80,34 @@ public FermaFlowPathRepository(FermaPersistentImplementation implementation) {
                 .map(FlowPath::new);
     }
 
+    @Override
+    public Map<PathId, FlowPath> findByIds(Set<PathId> pathIds) {
+        Set<String> graphPathIds = pathIds.stream()
+                .map(PathIdConverter.INSTANCE::toGraphProperty)
+                .collect(Collectors.toSet());
+        List<? extends FlowPathFrame> flowPathFrames = framedGraph().traverse(g -> g.V()
+                .hasLabel(FlowPathFrame.FRAME_LABEL)
+                .has(FlowPathFrame.PATH_ID_PROPERTY, P.within(graphPathIds)))
+                .toListExplicit(FlowPathFrame.class);
+        return flowPathFrames.stream()
+                .map(FlowPath::new)
+                .collect(Collectors.toMap(FlowPath::getPathId, Function.identity()));
+    }
+
+    @Override
+    public Map<PathId, Flow> findFlowsByPathIds(Set<PathId> pathIds) {
+        Set<String> graphPathIds = pathIds.stream()
+                .map(PathIdConverter.INSTANCE::toGraphProperty)
+                .collect(Collectors.toSet());
+        List<? extends FlowPathFrame> flowPathFrames = framedGraph().traverse(g -> g.V()
+                .hasLabel(FlowPathFrame.FRAME_LABEL)
+                .has(FlowPathFrame.PATH_ID_PROPERTY, P.within(graphPathIds)))
+                .toListExplicit(FlowPathFrame.class);
+        return flowPathFrames.stream()
+                .map(FlowPath::new)
+                .collect(Collectors.toMap(FlowPath::getPathId, FlowPath::getFlow));
```

Comment:
Not optimal. You can traverse further from FlowPath vertex (node) to Flow vertex, and get it.

Reply:
If I traverse further to Flow vertex I will loose selected pathIds so additional backward filtration will be required.